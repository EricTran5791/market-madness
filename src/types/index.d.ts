declare module '*.json' {
  /* tslint:disable no-any */
  const value: any;
  export default value;
}
