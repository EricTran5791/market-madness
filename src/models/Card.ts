import { types } from 'mobx-state-tree';
import {
  CardEffectUnion,
  CardEffectKind,
  BasicCardEffectSnapshotType,
  CardEffectCategory,
  InteractiveCardEffectSnapshotType,
  InteractiveCardEffectCategory,
} from './CardEffect';

export enum CardCategory {
  Action = 'Action',
  Consumable = 'Consumable',
  Item = 'Item',
  Money = 'Money',
  NPC = 'NPC',
}

export enum CardSubcategory {
  Food = 'Food',
}

export const Card = types
  .model({
    id: types.identifier(types.string),
    name: types.string,
    category: types.enumeration(
      'Category',
      Object.keys(CardCategory).map(key => CardCategory[key])
    ),
    subcategories: types.optional(
      types.array(
        types.enumeration(
          'Subcategory',
          Object.keys(CardSubcategory).map(key => CardSubcategory[key])
        )
      ),
      []
    ),
    description: types.optional(types.string, ''),
    cost: types.optional(types.number, 0),
    money: types.optional(types.number, 0),
    health: types.optional(types.number, 0),
    isPlayed: types.optional(types.boolean, false),
    effects: types.optional(types.array(CardEffectUnion), []),
  })
  .actions(self => ({
    afterCreate() {
      if (self.description.length > 0 || self.effects.length === 0) {
        return;
      }

      // Auto generate descriptions
      if (self.category === CardCategory.NPC) {
        self.description = 'Defeat: ';
      }
      self.description += self.effects
        .map(effect => {
          if (effect.kind === CardEffectKind.Basic) {
            const {
              category,
              value,
              gainedCardId,
            }: BasicCardEffectSnapshotType = effect;
            switch (category) {
              case CardEffectCategory.Draw:
                return `Draw ${value} card${value > 1 ? 's' : ''}`;
              case CardEffectCategory.GainAttack:
                return `+${value} Attack`;
              case CardEffectCategory.GainCardToDiscardPile:
                // The RegEx converts the camel case id to a spaced and capitalized name
                // TODO: There was a circular dependency when cardGenerator was imported
                // due to cardLibrary also importing CardCategory from Card
                return `Gain ${value} ${gainedCardId
                  .replace(/([A-Z])/g, ' $1')
                  .replace(/^./, (str: string) => {
                    return str.toUpperCase();
                  })}`;
              case CardEffectCategory.GainCardToHand:
                // The RegEx converts the camel case id to a spaced and capitalized name
                // TODO: There was a circular dependency when cardGenerator was imported
                // due to cardLibrary also importing CardCategory from Card
                return `Add ${value} ${gainedCardId
                  .replace(/([A-Z])/g, ' $1')
                  .replace(/^./, (str: string) => {
                    return str.toUpperCase();
                  })} to your hand`;
              case CardEffectCategory.GainMoney:
                return `+${value} Money`;
              case CardEffectCategory.Heal:
                return `Heal ${value}`;
              case CardEffectCategory.IncreaseMaxHealth:
                return `Increase max health by ${value}`;
              case CardEffectCategory.TrashSelf:
                return `Trash this card`;
              default:
                return '';
            }
          }
          if (effect.kind === CardEffectKind.Interactive) {
            const {
              category,
              numCardsToResolve,
            }: InteractiveCardEffectSnapshotType = effect;
            switch (category) {
              case InteractiveCardEffectCategory.OptionalTrash:
                return `Trash up to ${numCardsToResolve} cards`;
              case InteractiveCardEffectCategory.MandatoryDiscard:
                return `Discard ${numCardsToResolve} cards`;
              case InteractiveCardEffectCategory.MandatoryTrash:
                return `Trash ${numCardsToResolve} cards`;
              default:
                return '';
            }
          }
          return '';
        })
        .join(', ');
    },
    /** Reset a card's isPlayed status every time it is detached from its direct parent. */
    beforeDetach() {
      self.isPlayed = false;
    },
  }));

export type CardModelType = typeof Card.Type;
export type CardModelSnapshotType = typeof Card.SnapshotType;

export const CardStack = types
  .model({
    cards: types.array(Card),
  })
  .views(self => ({
    get totalCards(): number {
      return self.cards.length;
    },
    get totalMoney() {
      return self.cards
        .map(card => card.money)
        .reduce((sum, currentValue) => sum + currentValue, 0);
    },
    get unplayedCards() {
      return self.cards.filter(card => !card.isPlayed);
    },
  }))
  .actions(self => ({
    add(card: CardModelType) {
      self.cards.push(card);
    },
    remove(card: CardModelType) {
      self.cards.remove(card);
    },
  }));

export type CardStackModelType = typeof CardStack.Type;
