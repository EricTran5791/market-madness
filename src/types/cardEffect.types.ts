export enum CardEffectKind {
  Basic = 'Basic',
  Interactive = 'Interactive',
}

export enum CardEffectCategory {
  Draw = 'Draw',
  GainAttack = 'Gain Attack',
  GainCardToDiscardPile = 'Gain Card to Discard Pile',
  GainCardToHand = 'Gain Card to Hand',
  GainMoney = 'Gain Money',
  Heal = 'Heal',
  IncreaseMaxHealth = 'Increase Max Health',
  ShuffleCardToDeck = 'Shuffle Card to Deck',
  TopdeckCardToOpposingShopDecks = 'Topdeck Card to Opposing Shop Decks',
  TrashSelf = 'Trash Self',
}

export enum InteractiveCardEffectCategory {
  Discard = 'Discard',
  Trash = 'Trash',
}

export enum InteractiveCardEffectResolveType {
  Mandatory = 'Mandatory',
  Optional = 'Optional',
}

export type GainedCard = {
  id: string;
  name: string;
};

export interface BasicCardEffect {
  kind: CardEffectKind.Basic;
  category: CardEffectCategory;
  value: number;
  gainedCard?: GainedCard;
}

export interface InteractiveCardEffect {
  kind: CardEffectKind.Interactive;
  category: InteractiveCardEffectCategory;
  resolveType: InteractiveCardEffectResolveType;
  numPlaysToResolve?: number;
}

export type CardEffect = BasicCardEffect | InteractiveCardEffect;

export const initialBasicCardEffect: BasicCardEffect = {
  kind: CardEffectKind.Basic,
  category: CardEffectCategory.Draw,
  value: 0,
};

export const initialInteractiveCardEffect: InteractiveCardEffect = {
  kind: CardEffectKind.Interactive,
  category: InteractiveCardEffectCategory.Discard,
  resolveType: InteractiveCardEffectResolveType.Optional,
  numPlaysToResolve: 0,
};
