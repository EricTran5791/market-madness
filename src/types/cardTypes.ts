import {
  CardEffect,
  CardEffectKind,
  CardEffectCategory,
  InteractiveCardEffectCategory,
  InteractiveCardEffectResolveType,
} from './cardEffect.types';

export enum CardCategory {
  Action = 'Action',
  Consumable = 'Consumable',
  Item = 'Item',
  Money = 'Money',
  NPC = 'NPC',
}

export enum CardSubcategory {
  Food = 'Food',
  SportsEquipment = 'Sports Equipment',
  Tool = 'Tool',
}

export enum CardCostKind {
  Money = 'Money',
  Health = 'Health',
}

export type Card = {
  name: string;
  category: CardCategory;
  subcategories: CardSubcategory[];
  description: string;
  cost: {
    kind: CardCostKind;
    value: number;
  };
  effects: CardEffect[];
  isPlayed?: boolean;
};

export const initialCard: Card = {
  name: '',
  category: CardCategory.Item,
  subcategories: [],
  description: '',
  cost: {
    kind: CardCostKind.Money,
    value: 0,
  },
  // TEMP
  effects: [
    {
      category: CardEffectCategory.GainAttack,
      kind: CardEffectKind.Basic,
      value: 2,
    },
    {
      category: CardEffectCategory.Heal,
      kind: CardEffectKind.Basic,
      value: 2,
    },
    {
      category: InteractiveCardEffectCategory.Trash,
      kind: CardEffectKind.Interactive,
      resolveType: InteractiveCardEffectResolveType.Mandatory,
      numCardsToResolve: 2,
    },
  ],
};
