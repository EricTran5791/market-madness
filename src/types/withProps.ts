import { ThemedStyledFunction } from 'styled-components';

/** This is an unfortunate workaround to provide props to styled elements in Typescript:
 *  https://github.com/styled-components/styled-components/issues/630
 */
export const withProps = <U>() => <P, T, O>(
  fn: ThemedStyledFunction<P, T, O>
) => fn as ThemedStyledFunction<P & U, T, O & U>;
