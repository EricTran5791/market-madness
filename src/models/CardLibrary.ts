import { types, applySnapshot } from 'mobx-state-tree';
import { Card, CardModelSnapshotType, CardModelType } from './Card';
import { Map } from 'immutable';

export enum CardLibraryOperationKind {
  Success = 'Success',
  Error = 'Error',
}

export type CardLibraryOperation = {
  kind: CardLibraryOperationKind;
  text: string;
};

enum CardValidationStatusKind {
  Valid = 'Valid',
  Invalid = 'Invalid',
}

type CardValidationStatus = {
  kind: CardValidationStatusKind;
  text: string;
};

type ValidateCardParams = {
  card: CardModelType | CardModelSnapshotType;
  targetId: string;
};

export const CardLibrary = types
  .model('CardLibrary', {
    cards: types.optional(types.array(Card), []),
  })
  .views(self => ({
    /** Returns a map of the card library. Uses the card id as the key and the card view data as the value. */
    get cardLibraryJson() {
      return JSON.stringify(
        Map(self.cards.map(_ => [_.id, _.viewData])).toObject(),
        undefined,
        2
      );
    },
    getCardById(id: string): CardModelType {
      return self.cards.find(_ => _.id === id);
    },
  }))
  .actions(self => {
    function updateCard(
      id: string,
      snapshot: CardModelSnapshotType
    ): CardLibraryOperation {
      // Find the card to be updated
      const cardToUpdate = self.getCardById(id);
      const cardValidationStatus = validateCard({
        // Validate the snapshot
        card: snapshot,
        // If the id was changed in the snapshot, check that there isn't a card with the same id as the snapshot
        targetId: id !== snapshot.id ? snapshot.id : '',
      });

      if (
        cardToUpdate &&
        cardValidationStatus.kind === CardValidationStatusKind.Valid
      ) {
        applySnapshot(cardToUpdate, snapshot);
        return {
          kind: CardLibraryOperationKind.Success,
          text: `Success! ${
            snapshot.name
          } has been updated in the card library.`,
        };
      } else {
        return {
          kind: CardLibraryOperationKind.Error,
          text: cardValidationStatus.text,
        };
      }
    }

    function addCard(card: CardModelType): CardLibraryOperation {
      const cardValidationStatus = validateCard({ card, targetId: card.id });
      if (cardValidationStatus.kind === CardValidationStatusKind.Valid) {
        self.cards.push(card);
        return {
          kind: CardLibraryOperationKind.Success,
          text: `Success! ${card.name} has been added to the card library.`,
        };
      } else {
        return {
          kind: CardLibraryOperationKind.Error,
          text: cardValidationStatus.text,
        };
      }
    }

    function deleteCard(id: string) {
      const deletedCard = self.cards.find(_ => _.id === id);
      if (deletedCard) {
        self.cards.remove(deletedCard);
      }
    }

    function validateCard({
      card,
      targetId,
    }: ValidateCardParams): CardValidationStatus {
      // Check to see if there is another card in the library with the same id
      if (targetId && self.getCardById(targetId)) {
        return {
          kind: CardValidationStatusKind.Invalid,
          text:
            'There is another card in the library with the same id. Please use another name.',
        };
      }

      // Check for an empty id or name
      if (card.id === '' || card.name === '') {
        return {
          kind: CardValidationStatusKind.Invalid,
          text:
            'Error! The card id and name must not be empty. Please change the card name.',
        };
      }

      return { kind: CardValidationStatusKind.Valid, text: '' };
    }

    return {
      updateCard,
      addCard,
      deleteCard,
      validateCard,
    };
  });

export type CardLibraryModelType = typeof CardLibrary.Type;
