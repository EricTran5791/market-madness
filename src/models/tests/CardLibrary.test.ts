import { CardLibraryModelType, CardLibrary } from '../CardLibrary';
import { getSnapshot } from 'mobx-state-tree';
import { Card } from '../Card';
import { initialCardState } from '../../types/cardTypes';
import { initialBasicCardEffect } from '../../types/cardEffect.types';

describe('Card Library', () => {
  let mockCardLibrary: CardLibraryModelType;
  const mockCard = {
    ...initialCardState,
    uniqid: '1',
    id: '1',
    name: 'Card 1',
  };

  beforeEach(() => {
    mockCardLibrary = CardLibrary.create({
      cards: [
        Card.create({ ...mockCard }),
        Card.create({
          ...initialCardState,
          uniqid: '2',
          id: '2',
          name: 'Card 2',
        }),
        Card.create({
          ...initialCardState,
          uniqid: '3',
          id: '3',
          name: 'Card 3',
        }),
      ],
    });
  });

  it('creates a Card Library', () => {
    expect(getSnapshot(mockCardLibrary)).toMatchSnapshot();
  });

  it('gets a card by id from the library', () => {
    expect(mockCardLibrary.getCardById('1')).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('2')).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('undefined')).toMatchSnapshot();
  });

  it('updates a card in the library', () => {
    expect(
      mockCardLibrary.updateCard('1', { ...mockCard, name: 'Card 1 New' })
    ).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('1')).toMatchSnapshot();
  });

  it('fails to update a card in the library when the snapshot id is being used by another card in the library', () => {
    expect(
      mockCardLibrary.updateCard('1', {
        ...mockCard,
        id: '2',
        name: 'Card 1 New',
      })
    ).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('1')).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('2')).toMatchSnapshot();
  });

  it('adds a new card to the library', () => {
    expect(
      mockCardLibrary.addCard(
        Card.create({
          ...initialCardState,
          uniqid: 'new',
          id: 'new',
          name: 'New Card',
        })
      )
    ).toMatchSnapshot();
  });

  it(`fails to add a new card to the library when the new card id
    is already being used by another card in the library`, () => {
    expect(
      mockCardLibrary.addCard(
        Card.create({
          ...initialCardState,
          uniqid: 'new',
          id: '1',
          name: 'New Card',
        })
      )
    ).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('1')).toMatchSnapshot();
  });

  it('fails to add a new card to the library when the new card id or name is empty', () => {
    expect(
      mockCardLibrary.addCard(
        Card.create({
          ...initialCardState,
          uniqid: 'new',
          id: '',
          name: 'New Card',
        })
      )
    ).toMatchSnapshot();
    expect(
      mockCardLibrary.addCard(
        Card.create({
          ...initialCardState,
          uniqid: 'new',
          id: 'new',
          name: '',
        })
      )
    ).toMatchSnapshot();
  });

  it('deletes a card in the library', () => {
    expect(mockCardLibrary.deleteCard('1')).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('1')).toMatchSnapshot();
  });

  it('fails to delete an non-existent card in the library', () => {
    expect(mockCardLibrary.deleteCard('undefined')).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('undefined')).toMatchSnapshot();
  });

  it('fails to delete a card that is being referenced by another card in the library', () => {
    mockCardLibrary.addCard(
      Card.create({
        ...initialCardState,
        uniqid: 'new',
        id: 'new',
        name: 'New Card that references Card 1',
        effects: [
          { ...initialBasicCardEffect, gainedCard: { id: '1', name: '1' } },
        ],
      })
    );
    expect(mockCardLibrary.deleteCard('1')).toMatchSnapshot();
    expect(mockCardLibrary.getCardById('1')).toMatchSnapshot();
  });
});
