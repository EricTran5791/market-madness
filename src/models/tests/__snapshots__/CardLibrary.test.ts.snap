// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Library adds a new card to the library 1`] = `
Object {
  "kind": "Success",
  "text": "Success! New Card has been added to the card library.",
}
`;

exports[`Card Library creates a Card Library 1`] = `
Object {
  "cardFilter": Object {
    "filterText": "",
    "sortMode": "Alphabetical Ascending",
  },
  "cards": Array [
    Object {
      "category": "Neutral",
      "cost": Object {
        "kind": "Money",
        "value": 0,
      },
      "description": "",
      "effects": Array [],
      "id": "1",
      "isPlayed": false,
      "name": "Card 1",
      "subcategories": Array [],
      "uniqid": "1",
    },
    Object {
      "category": "Neutral",
      "cost": Object {
        "kind": "Money",
        "value": 0,
      },
      "description": "",
      "effects": Array [],
      "id": "2",
      "isPlayed": false,
      "name": "Card 2",
      "subcategories": Array [],
      "uniqid": "2",
    },
    Object {
      "category": "Neutral",
      "cost": Object {
        "kind": "Money",
        "value": 0,
      },
      "description": "",
      "effects": Array [],
      "id": "3",
      "isPlayed": false,
      "name": "Card 3",
      "subcategories": Array [],
      "uniqid": "3",
    },
  ],
}
`;

exports[`Card Library deletes a card in the library 1`] = `
Object {
  "kind": "Success",
  "text": "",
}
`;

exports[`Card Library deletes a card in the library 2`] = `undefined`;

exports[`Card Library fails to add a new card to the library when the new card id
    is already being used by another card in the library 1`] = `
Object {
  "kind": "Error",
  "text": "There is another card in the library with the same id. Please use another name.",
}
`;

exports[`Card Library fails to add a new card to the library when the new card id
    is already being used by another card in the library 2`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "1",
  "isPlayed": false,
  "name": "Card 1",
  "subcategories": Array [],
  "uniqid": "1",
}
`;

exports[`Card Library fails to add a new card to the library when the new card id or name is empty 1`] = `
Object {
  "kind": "Error",
  "text": "Error! The card id and name must not be empty. Please change the card name.",
}
`;

exports[`Card Library fails to add a new card to the library when the new card id or name is empty 2`] = `
Object {
  "kind": "Error",
  "text": "Error! The card id and name must not be empty. Please change the card name.",
}
`;

exports[`Card Library fails to delete a card that is being referenced by another card in the library 1`] = `
Object {
  "kind": "Error",
  "text": "Error! Unable to delete Card 1, it is still referenced by the following
            cards: New Card that references Card 1.",
}
`;

exports[`Card Library fails to delete a card that is being referenced by another card in the library 2`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "1",
  "isPlayed": false,
  "name": "Card 1",
  "subcategories": Array [],
  "uniqid": "1",
}
`;

exports[`Card Library fails to delete an non-existent card in the library 1`] = `
Object {
  "kind": "Error",
  "text": "Error! Card deletion was unsuccessful because it does not exist in the card library.",
}
`;

exports[`Card Library fails to delete an non-existent card in the library 2`] = `undefined`;

exports[`Card Library fails to update a card in the library when the snapshot id is being used by another card in the library 1`] = `
Object {
  "kind": "Error",
  "text": "There is another card in the library with the same id. Please use another name.",
}
`;

exports[`Card Library fails to update a card in the library when the snapshot id is being used by another card in the library 2`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "1",
  "isPlayed": false,
  "name": "Card 1",
  "subcategories": Array [],
  "uniqid": "1",
}
`;

exports[`Card Library fails to update a card in the library when the snapshot id is being used by another card in the library 3`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "2",
  "isPlayed": false,
  "name": "Card 2",
  "subcategories": Array [],
  "uniqid": "2",
}
`;

exports[`Card Library gets a card by id from the library 1`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "1",
  "isPlayed": false,
  "name": "Card 1",
  "subcategories": Array [],
  "uniqid": "1",
}
`;

exports[`Card Library gets a card by id from the library 2`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "2",
  "isPlayed": false,
  "name": "Card 2",
  "subcategories": Array [],
  "uniqid": "2",
}
`;

exports[`Card Library gets a card by id from the library 3`] = `undefined`;

exports[`Card Library updates a card in the library 1`] = `
Object {
  "kind": "Success",
  "text": "Success! Card 1 New has been updated in the card library.",
}
`;

exports[`Card Library updates a card in the library 2`] = `
Object {
  "category": "Neutral",
  "cost": Object {
    "kind": "Money",
    "value": 0,
  },
  "description": "",
  "effects": Array [],
  "id": "1",
  "isPlayed": false,
  "name": "Card 1 New",
  "subcategories": Array [],
  "uniqid": "1",
}
`;
